ddtrace/contrib/cassandra/session.py

```diff
@@ -31,32 +36,133 @@ def traced_connect(func, instance, args, kwargs):
 
-    with tracer.trace("cassandra.query", service=service, span_type=cassx.TYPE) as span:
-        _sanitize_query(span, query)
-        span.set_tags(_extract_session_metas(instance))     # FIXME[matt] do once?
-        span.set_tags(_extract_cluster_metas(cluster))
-        result = None
-        try:
-            result = func(*args, **kwargs)
-            return result
-        finally:
-            if result:
-                span.set_tags(_extract_result_metas(result))
+    span = _start_span_and_set_tags(pin, query, session, cluster, page_number)
+
+    setattr(instance, PAGE_NUMBER, page_number)
+    setattr(instance, CURRENT_SPAN, span)
+    try:
+        return func(*args, **kwargs)
+    except:
+        span.set_exc_info(*sys.exc_info())
+        span.finish()
+        raise
 
+def traced_execute_async(func, instance, args, kwargs):
+    cluster = getattr(instance, 'cluster', None)
+    pin = Pin.get_from(cluster)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    query = kwargs.get("query") or args[0]
+
+    span = _start_span_and_set_tags(pin, query, instance, cluster, page_number=1)
+
+    try:
+        result = func(*args, **kwargs)
+        setattr(result, CURRENT_SPAN, span)
+        setattr(result, PAGE_NUMBER, 1)
+        setattr(
+            result,
+            '_set_final_result',
+            wrapt.FunctionWrapper(
+                result._set_final_result,
+                traced_set_final_result
+            )
+        )
+        setattr(
+            result,
+            '_set_final_exception',
+            wrapt.FunctionWrapper(
+                result._set_final_exception,
+                traced_set_final_exception
+            )
+        )
+        setattr(
+            result,
+            'start_fetching_next_page',
+            wrapt.FunctionWrapper(
+                result.start_fetching_next_page,
+                traced_start_fetching_next_page
+            )
+        )
+        # Since we cannot be sure that the previous methods were overwritten
+        # before the call ended, we add callbacks that will be run
+        # synchronously if the call already returned and we remove them right
+        # after.
+        result.add_callbacks(
+            _close_span_on_success,
+            _close_span_on_error,
+            callback_args=(result,),
+            errback_args=(result,)
+        )
+        result.clear_callbacks()
+        return result
+    except:
+        span.set_exc_info(*sys.exc_info())
+        span.finish()
+        raise
+
+def _start_span_and_set_tags(pin, query, session, cluster, page_number):
+    service = pin.service
+    tracer = pin.tracer
+    span = tracer.trace("cassandra.query", service=service, span_type=cassx.TYPE)
+    _sanitize_query(span, query)
+    span.set_tags(_extract_session_metas(session))     # FIXME[matt] do once?
+    span.set_tags(_extract_cluster_metas(cluster))
+    span.set_tag(cassx.PAGE_NUMBER, page_number)
```

Comment:
do we always have this tag? because if the response is not paginated, I'd say not to add this tag.